### PR TITLE
added repository dispatch trigger in github workflow to allow schematic to trigger DCA deployment 

### PR DESCRIPTION
**Pull request by linglp**
Related to the issue here: https://github.com/Sage-Bionetworks/schematic/issues/1026
Hopefully could give it a try :)

### CODE REVIEW

**Comment by rrchai** [comment]
@linglp Thanks for testing the repo dispatch. It looks like the deployment failed, but feel free to re-run the workflow - it might be just one-time issue.

Since the destination of deployment in our current workflow file is still depended on the branch name, two places are also needed to be considered:
- [here](https://github.com/Sage-Bionetworks/data_curator/blob/e13593c27378f4c02db1d0d448ab57c45c4f53c2/.github/workflows/shinyapps_deploy.yml#L51-L67) is where the `main/develop` branch of `schematic` is going to clone
- [here](https://github.com/Sage-Bionetworks/data_curator/blob/e13593c27378f4c02db1d0d448ab57c45c4f53c2/.github/workflows/shinyapps_deploy.yml#L103-L112) is codes to determine which instance (prod/staging//testing1) is being deployed. The develop branch is only 

I think using repo dispatch to trigger opening a PR might be a way, like:
the request will be sent when the `develop` branch of `schematic` updates -> open a pr based on `develop` of dca -> opening a PR will auto trigger the deployment to `testing1` instance which will use `develop` branch of `schematic`

**Comment by linglp** [comment]
> @linglp Thanks for testing the repo dispatch. It looks like the deployment failed, but feel free to re-run the workflow - it might be just one-time issue.
> 
> Since the destination of deployment in our current workflow file is still depended on the branch name, two places are also needed to be considered:
> 
> * [here](https://github.com/Sage-Bionetworks/data_curator/blob/e13593c27378f4c02db1d0d448ab57c45c4f53c2/.github/workflows/shinyapps_deploy.yml#L51-L67) is where the `main/develop` branch of `schematic` is going to clone
> * [here](https://github.com/Sage-Bionetworks/data_curator/blob/e13593c27378f4c02db1d0d448ab57c45c4f53c2/.github/workflows/shinyapps_deploy.yml#L103-L112) is codes to determine which instance (prod/staging/testing1) is being deployed.
> 
> I think using repo dispatch to trigger opening a PR might be an alternative way (also developers can review if the latest develop of schematic passes the workflow), it will be like:
> 
> 1. the request will be sent when the `develop` branch of `schematic` updates
> 2. open a pr compared to `develop` of dca
> 3. opening a PR will auto trigger the deployment to [testing1](https://sagebio.shinyapps.io/data_curator-testing1/) instance which will use `develop` branch of `schematic`

Thanks @rrchai ! I tried running the workflow again and it worked. Could you allow github action to trigger opening PRs? I don't have the permission to do it. You could do it via repository's settings: Actions > General > Workflow permissions.

**Comment by linglp** [comment]
Related to PR from schematic: https://github.com/Sage-Bionetworks/schematic/pull/1044

File: .github/workflows/open-PR.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      - name: Print out branch info
+        run: 'echo "Triggered by event branch: ${{ github.event.client_payload.branch }}"'
+        
+      - name: Check out repository
+        uses: actions/checkout@v3
+        with:
+          ref: develop
+
+      - name: Make mock changes to pull request
+        run: date +%s > mock.txt
```
Comment (rrchai) [inline .github/workflows/open-PR.yml:20]: I suggest to update another file, like `schematic_config.yml`

**Review by rrchai — CHANGES_REQUESTED** [review]

**Comment by rrchai** [comment]
Thanks @linglp. I would suggest to add one property ("version") in `schematic_config.yml` and update the version each event. It can be also used to update from `release` changes from schematic as well. 

I made the attempt with [this wl](https://github.com/rrchai/rong-flow-test/blob/main/.github/workflows/ci.yaml) using [yaml-update-action](https://github.com/fjogeleit/yaml-update-action), which can update the value of yaml file and create a pr. The example of created PR is [here](https://github.com/rrchai/rong-flow-test/pull/4).

**Comment by linglp** [comment]
> Thanks @linglp. Instead of updating `report.txt`, I would suggest to update `schematic_config.yml` (you can add "version" property in the file). So the version of schematic used is stored in the config file, which we could use for the app. Also, it can be used for `release` updates of schematic - just need to replace with the version tag.
> 
> I made the attempt with [this wl](https://github.com/rrchai/rong-flow-test/blob/main/.github/workflows/ci.yaml) using [yaml-update-action](https://github.com/fjogeleit/yaml-update-action), which can update the value of yaml file and create a pr. The example of created PR is [here](https://github.com/rrchai/rong-flow-test/pull/4). Let me know if you have any questions.

Thanks @rrchai ! Your suggestion makes sense. I have updated the PR to use the workflows that you recommended. One thing that I am not sure is how to "switch" between "develop" and "main" branch based on if the release tag exists or not by using the `targetBranch` parameter. Let me know if you have any thought.

File: .github/workflows/open-PR.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        run: |
+          echo 'COMMIT_REF=${{ github.event.client_payload.commit-sha }}' >> $GITHUB_ENV
+          echo 'BRANCH_NAME=${{ github.event.client_payload.branch }}' >> $GITHUB_ENV
+          echo 'RELEASE_VERSION=${{ github.event.client_payload.release-version }}' >> $GITHUB_ENV
+          echo 'DEFAULT_VAL='empty'' >> $GITHUB_ENV
+
+      # create different pull request based on if release tag exists. 
+      - name: Update values.yaml
+        uses: fjogeleit/yaml-update-action@main
+        if: ${{ env.RELEASE_VERSION != '' }}
```
Comment (linglp) [inline .github/workflows/open-PR.yml:34]: if release tag does not exist, trigger a PR to merge into the `develop` branch

File: .github/workflows/open-PR.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            {
+              "schematic.branch": "${{ env.BRANCH_NAME|| env.DEFAULT_VAL }}",
+              "schematic.sha": "${{ env.COMMMIT_REF || env.DEFAULT_VAL }}",
+              "schematic.release_version": "${{ env.RELEASE_VERSION || env.DEFAULT_VAL}}"
+            }
+
+      # create pull request
+      - name: Update values.yaml
+        uses: fjogeleit/yaml-update-action@main
+        if: ${{ env.RELEASE_VERSION == '' }}
```
Comment (linglp) [inline .github/workflows/open-PR.yml:55]: if the release tag does not exist, trigger a PR to update the develop branch

File: .github/workflows/open-PR.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      #### create different pull request based on if release tag exists. 
+      ## when a new release is available
+      - name: Update values.yaml
+        uses: fjogeleit/yaml-update-action@main
+        if: ${{ env.RELEASE_VERSION != '' }}
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          valueFile: schematic_config.yml
+          branch: develop-update-schematic
+          targetBranch: main
```
Comment (linglp) [inline .github/workflows/open-PR.yml:36]: change to "targetBranch: develop"

File: .github/workflows/open-PR.yml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          labels: 'auto-release'
+          title: 'Version Updates triggered by schematic develop branch'
+          description: Triggered by schematic updates 
+          message: 'Triggered by schematic updates'
+          changes: |
+            {
+              "schematic.branch": "develop",
+              "schematic.sha": "${{ env.COMMMIT_REF || env.DEFAULT_VAL }}",
+              "schematic.release_version": "${{env.DEFAULT_VAL}}"
+            }
```
Comment (rrchai) [inline .github/workflows/open-PR.yml:68]: remove this part - duplicated

**Review by rrchai — APPROVED** [review]
LGTM! Just one comment

**Comment by rrchai** [comment]
Merging this PR to `develop`

**Pull request merged**